cdap-app-templates/cdap-etl/cdap-data-pipeline/src/test/java/co/cask/cdap/datapipeline/DataPipelineTest.java

```diff
@@ -1771,4 +1778,102 @@ private void validateMetric(long expected, ApplicationId appId,
     // wait for won't throw an exception if the metric count is greater than expected
     Assert.assertEquals(expected, getMetricsManager().getTotalMetric(tags, "user." + metric));
   }
+
+  @Test
+  public void testServiceUrlMR() throws Exception {
+    testServiceUrl(Engine.MAPREDUCE);
+  }
+
+  @Test
+  public void testServiceUrlSpark() throws Exception {
+    testServiceUrl(Engine.SPARK);
+  }
+
+  public void testServiceUrl(Engine engine) throws Exception {
+
+    // Deploy the ServiceApp application
+    ApplicationManager appManager = deployApplication(ServiceApp.class);
+
+    // Start WhoFlow
+    FlowManager flowManager = appManager.getFlowManager("WhoFlow").start();
```

Comment:
instead of running a flow that reads from a stream and writes to the dataset, it would be faster and simpler to just add another service endpoint that writes a value from the request body into the dataset.